public/app/features/dashboard/dashgrid/PanelPerformanceMonitor.tsx

```diff
@@ -0,0 +1,163 @@
+import appEvents from 'app/core/app_events';
+import { config } from 'app/core/config';
+import { FIELD_CONFIG_CUSTOM_KEY, FIELD_CONFIG_OVERRIDES_KEY, PanelLogEvents } from 'app/core/log_events';
+import { DashboardSavedEvent } from 'app/types/events';
+
+interface Props {
+  isInPanelEdit: boolean;
+  panelType: string;
+  panelId: number;
+  panelTitle: string;
+  panelOptions: any;
+  panelFieldConfig: FieldConfigSource;
+}
+
+export const PanelPerformanceMonitor = (props: Props) => {
+  const startLoadTime = performance.now();
+  const panelIdStr = String(props.panelId);
+
+  const oldPanelOptions = useRef(props.panelOptions);
+  const oldPanelFieldConfig = useRef<FieldConfigSource>(props.panelFieldConfig);
+
+  const savedFieldConfig = useRef<FieldConfigSource>(props.panelFieldConfig);
+  const savedPanelOptions = useRef(props.panelOptions);
+
+  const logChanges = () => {
+    logPanelOptionChanges(savedPanelOptions.current, oldPanelOptions.current);
+    logFieldConfigChanges(savedFieldConfig.current, oldPanelFieldConfig.current);
+
+    //set the old values to the current values for next save
+    oldPanelOptions.current = savedPanelOptions.current;
+    oldPanelFieldConfig.current = savedFieldConfig.current;
+  };
+
+  useEffect(() => {
+    savedPanelOptions.current = props.panelOptions;
+  }, [props.panelOptions]);
+
+  useEffect(() => {
+    savedFieldConfig.current = props.panelFieldConfig;
+  }, [props.panelFieldConfig]);
+
+  const logEvent = (eventName: string, newKey: string, newVal: string, oldVal?: string) => {
+    const logObj = {
+      key: newKey,
+      newValue: newVal,
+      oldValue: oldVal ?? '',
+      panelTitle: props.panelTitle,
+      panelId: panelIdStr,
+      panelType: props.panelType,
+    };
+
+    faro.api.pushEvent(eventName, logObj);
+  };
+
+  const logPanelOptionChanges = (panelOptions: any, oldPanelOptions: any) => {
+    for (const [key, value] of Object.entries(panelOptions)) {
+      const newValue: string = typeof value !== 'string' ? JSON.stringify(value) : value;
+      const oldValue: string = typeof value !== 'string' ? JSON.stringify(oldPanelOptions[key]) : oldPanelOptions[key];
+
+      if (oldPanelOptions[key] === undefined) {
+        logEvent(PanelLogEvents.NEW_PANEL_OPTION_EVENT, key, newValue);
+      } else if (oldValue !== newValue) {
+        logEvent(PanelLogEvents.PANEL_OPTION_CHANGED_EVENT, key, newValue, oldValue);
+      }
+    }
+  };
+
+  const logFieldConfigChanges = (fieldConfig: FieldConfigSource, oldFieldConfig: FieldConfigSource) => {
+    // overrides are an array of objects, so stringify it all and log changes
+    // in lack of an index, we can't tell which override changed
+    if (oldFieldConfig.overrides !== fieldConfig.overrides) {
+      logEvent(
+        PanelLogEvents.FIELD_CONFIG_OVERRIDES_CHANGED_EVENT,
+        FIELD_CONFIG_OVERRIDES_KEY,
+        JSON.stringify(fieldConfig.overrides),
+        JSON.stringify(oldFieldConfig.overrides)
+      );
+    }
```

Comment:
I'd like to get understanding how field config / options changes information will be used. Seems like logging those is a lot. How would those be correlated with time perfromance metrics - espacially that the only perf metric I see in this PR is panel load time.

Reply:
@mdvictor - if your point of logging those is to get an understanding of what options and field config options are used, I think you may want to talk with @thanos-karachalios and @shih-chris, they are working on https://docs.google.com/document/d/19eTPjmXMlvBpxQEQynCntfM9A5RSqIV9ont_s3QjfI8/

Reply:
Some dashboard stats are already reported in https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/utils/tracking.ts#L15, think we could add some panel reports there too. Maybe not to the dashboard event, but to the same code pipeline. IMO gathering options/field config information on dashboard load should be sufficient for usage analysis. The options/field config change doesn't seem like bringing a lot of value.

Reply:
Thank you @dprokop. The purpose of these logs and measurements isn't to track usage necessarily but to offer us better observability of our panels health in the context of our [frontend monitoring initiative](https://docs.google.com/document/d/1DeXnMOyVFc_OIIlC16WMFy-zEcPqKkrGE9DSYWq6SQ0/) . Faro gives us some OOTB metrics/logs and these would supplement them for a better understanding.